src/content_stream.cpp

```diff
@@ -43,118 +43,164 @@ void ContentReader::close() {
   }
 }
 
-std::string ContentReader::read() {
-  matched_idx = 0;
+// Limit message sizes to 10 MB to prevent DoS issues
+#define MAX_CONTENT_LENGTH (100000000)
 
-  // Find Content-Length header prefix
-  if (!scan("Content-Length:")) {
-    return "";
+std::string ContentReader::read() {
+  // Currently, a DAP message is encoded like:
+  // Content-Length: 2\r\n
+  // \r\n
+  // {}
+  auto contentLengthResult = readContentLength();
+  if (!contentLengthResult.first) {
+    return invalidMsg();
   }
 
-  // Skip whitespace and tabs
-  while (matchAny(" \t")) {
-  }
-
-  // Parse length
-  size_t len = 0;
-  while (true) {
-    auto c = matchAny("0123456789");
-    if (c == 0) {
-      break;
-    }
-    len *= 10;
-    len += size_t(c) - size_t('0');
-  }
-  if (len == 0) {
-    return "";
-  }
-  // Expect \r\n\r\n
-  if (!match("\r\n\r\n")) {
-    return "";
-  }
+  auto len = contentLengthResult.second;
 
   // Read message
-  if (!buffer(len + matched_idx)) {
-    return "";
-  }
+  std::string out;
+  out.reserve(len);
 
-  for (size_t i = 0; i < matched_idx; i++) {
+  // Start with any extra data read when parsing the header
+  while (out.size() < len && !buf.empty()) {
+    uint8_t c = buf.front();
+    out.push_back(static_cast<char>(c));
     buf.pop_front();
   }
 
-  std::string out;
-  out.reserve(len);
-  for (size_t i = 0; i < len; i++) {
-    out.push_back(static_cast<char>(buf.front()));
-    buf.pop_front();
+  if (!readContent(out, len)) {
+    return invalidMsg();
   }
+
   return out;
 }
 
-bool ContentReader::scan(const uint8_t* seq, size_t len) {
-  while (buffer(len)) {
-    if (match(seq, len)) {
-      return true;
-    }
-    buf.pop_front();
-  }
-  return false;
+std::string ContentReader::invalidMsg() {
+  this->close();
+  return "";
 }
 
-bool ContentReader::scan(const char* str) {
-  auto len = strlen(str);
-  return scan(reinterpret_cast<const uint8_t*>(str), len);
+// Given a max content length of 10 MB, the Content-Length header field
+// has a known max length:
+// (the length of "Content-Length") +
+// (the length of the max content length value) +
+// (the length of "\r\n")*2
+#define MAX_HEADER_LENGTH (16 + 9 + 2 + 2)
+
+static inline std::pair<bool, size_t> invalidHeader() {
+  return std::make_tuple(false, 0);
 }
 
-bool ContentReader::match(const uint8_t* seq, size_t len) {
-  if (!buffer(len + matched_idx)) {
-    return false;
-  }
-  auto it = matched_idx;
-  for (size_t i = 0; i < len; i++, it++) {
-    if (buf[it] != seq[i]) {
-      return false;
+std::pair<bool, size_t> ContentReader::readContentLength() {
+  std::string header;
+  header.reserve(MAX_HEADER_LENGTH);
```

Comment:
This is likely to incur a heap allocation in very hot code. Given that this is a known maximum size, can we store this on the stack?